### PR TITLE
More LBF Updates

### PR DESCRIPTION
**Pull request by sjpadgett**
1. Replace HTML2PDF with mPDF.
2. More restyle(mainly menu nav bar and add group)
3. Add four new data type modifiers for styling.
4. Tried to get stacking order in check!
5. Minor bug fixes.
6. New usage Tips panel.

(below added by bradymiller)
Demo is here:
http://www.open-emr.org/wiki/index.php/Development_Demo#Mu_-_Up_For_Grabs_Demo

### CODE REVIEW

File: interface/super/edit_layout.php [hunk 1]
```diff
@@ -1600,7 +1632,13 @@ function myChangeCheck() {
 	{id: 'U',text:'" . xla('Capitalize all') . "'},
 	{id: 'V',text:'" . xla('Vendor') . "'},
 	{id: 'X',text:'" . xla('Do Not Print') . "'},
-	{id: '0',text:'" . xla('Read Only') . "'},
+    {id:'grp',text:'" . xla('Stylings') . "',children:[
+        {id: 'RS',text:'" . xla('Add Bottom Border Row') . "'},
+        {id: 'RO',text:'" . xla('Outline Entire Row') . "'},
+        {id: 'DS',text:'" . xla('Add Data Bottom Border') . "'},
+        {id: 'DO',text:'" . xla('Outline Data Col') . "'}
+    ]},
```
Comment (bradymiller) [inline interface/super/edit_layout.php]: Nice options. As I recall, somebody was asking how to do this recently on forums or chat.

Comment (bradymiller) [reply interface/super/edit_layout.php]: To clarify, are 2 digit options supported. For example, will this also turn on the 'D' and 'o' and 'R' and 'S' options when used downstream.

**Comment by sunsetsystems** [comment]
It's nice to see someone besides me improving LBF stuff. I'll try to find some time today or tomorrow to look it over.

**Comment by bradymiller** [comment]
Can demo this here:
http://www.open-emr.org/wiki/index.php/Development_Demo#Mu_-_Up_For_Grabs_Demo

**Comment by sjpadgett** [comment]
Rod, I now realize just how useful this feature is. I still have data type context nagging at me and I'd love to see a dialog/screen that maps the layout on a box grid where you could drag and drop items in a wysiwyg fashion. Each box would have properties for that item etc.

File: interface/super/edit_layout.php (hunk 1 above)
Comment (sjpadgett) [reply interface/super/edit_layout.php]: It's the reason I set these up as json. But nowww you have me wary, I better check again:)

Comment (sjpadgett) [reply interface/super/edit_layout.php]: Sure enough, guess I have to get creative...

**Comment by sjpadgett** [comment]
Thanks for catch @bradymiller . Here is my solution. Also, any tips to add to new tip panel please let me know. Think I will open forum discussion to garner more maybe.

File: interface/super/edit_layout.php [hunk 2]
```diff
@@ -1271,21 +1281,41 @@ function myChangeCheck() {
...
+</span>
 <?php } else { ?>
 <input type='button' value='<?php echo xla('New Layout'); ?>' onclick='edit_layout_props("")' />&nbsp;
 <?php } ?>
-</p></div>
-</div>
 
+<div id="tips" class="container tips"><section class="panel panel-primary">
+  <header class="panel-heading">
+   <h3 class="panel-title">Usage Tips</h3>
```
Comment (bradymiller) [inline interface/super/edit_layout.php]: After substantial digging here, I have finally found something that needs to be translated :)

Comment (sjpadgett) [reply interface/super/edit_layout.php]: Honestly my friend, I truly do not have anything against translating for our non English speaking brothers and sisters:) I am however a little put off we don't translate southerner!

File: library/options.inc.php [hunk 3]
```diff
@@ -3966,4 +3966,28 @@ function lbf_canvas_head($small = true)
     return $s;
 }
 
+/**
+ * Test if modifier($test) is in array of options for data type.
+ *
+ * @param json array $options or could be string of form "ABCU"
```
Comment (bradymiller) [inline library/options.inc.php]: Very smart to support both!

File: library/options.inc.php [hunk 4]
```diff
@@ -3966,4 +3966,28 @@ function lbf_canvas_head($small = true)
...
+    if (strpos($options, ',') === false) { // could be string of char's or single element of json
+        json_decode($options);
+        if (is_string($options) && ! (json_last_error() === JSON_ERROR_NONE)) { // nope, it's string.
+            $t = str_split(trim($options)); // very good chance it's legacy modifier string.
+            $options = json_encode($t); // make it array.
+        }
+    }
+    $options = json_decode($options);
+    
+    return in_array($test, $options, true) ? true : false; // finally!
```
Comment (bradymiller) [inline library/options.inc.php]: Case in point. Who needs their own coding style as a creative outlet when we have comments :)

**Comment by bradymiller** [comment]
Code looks good to me. If this is testing well for you, just let me know; and then I'll bring it into the codebase.

File: library/options.inc.php (hunk 4 above)
Comment (sjpadgett) [reply library/options.inc.php]: Don't encourage me too much or my code will turn into a book :). I think I may add one more modifier or data type for paragraph. Mainly for print. Static text is fine but getting the text to span entire row is challenge. Will let ya know.

**Comment by sjpadgett** [comment]
Okay, I'm done for now but have plans to open new PR soon to add context and I am considering a modal to show the encounter edit for a reference of how form looks while building. Also I found a really nice form creator jquery api for drag and drop that is very extendable. Don't see it being very hard to use for our purpose. See https://formbuilder.online/. And for gods sake, someone stop me!!!!!

**Comment by sjpadgett** [comment]
@bradymiller I think we need to add jquery migrate lib to header call(yaml). This will allow using the header class everywhere plus identify depreciated items that have been removed up till v3.0.0, hopefully prompting folks while working in code to fix depreciation.
I fixed eModal to be compatible with current jquery versions. ( i like eModal and think we can take it over, so easy to use). I also added a popup showing the encounter version of current lbf form being edited which makes building form somewhat easier. Not sure if I should commit here or a new PR.
I hope I'm not bugging y'all too much, I seem to not be able to stop myself when this feature is so useful.
Jquery migrate is at: https://github.com/jquery/jquery-migrate/

**Comment by sjpadgett** [comment]
Okay Brady I promise i'm finished. If you haven't any issues then this is testing well for me. The encounter preview makes it so much easier to create/test forms. Fixed eModal so conforms to current jquery version and plan to fork and modify for OpenEMR. Very handy tool.
Sorry so many post but i'm a sharing kind of guy. Thanks.

**Comment by bradymiller** [comment]
very nice! The form preview thing is a really nice addition!
Just brought your code into the codebase. I  followed it with a commit that brings emodal into the yaml config file.
The goal is to never edit assets, but didn't see around it in this situation, so place the following on the bower wiki page to fix this at some point (either by bringing in a new version if that happens or forking it and pulling our fork in instead):
http://www.open-emr.org/wiki/index.php/Bower#emodal-1-2-65

**Comment by sjpadgett** [comment]
Yep I get custom assets. A chance the mod gets lost. I will follow up with a solution of either getting the awol dev of eModal to update ,fork it or write my own. Just easier to write five lines of code for five modals than create 5 divs.

**Pull request merged**